KAFKA-7523: Add ConnectedStoreProvider to Processor API

pgwhalen opened the pull request:
- Implementing KIP-401, ConnectedStoreProvider is now the
 superinterface of Processor/TransformerSuppliers, allowing the user to
 provide state stores that will automatically be added to the topology
 and connected to the associated processor or transformer.  This is an
 alternate way to connect state stores to a processor, and cannot be
 used along with passing the varags stateStoreNames to
 KStream::process/transform. It also works with Topology::addProcessor.
 - WordCountTransformerDemo is a middle ground between the low level
 Processor API and the high level DSL, using KStream::transform and
 ConnectedStoreProvider
 - The restriction on adding the same state store twice is relaxed so
 that multiple ConnectedStoreProviders can add the same store.

https://issues.apache.org/jira/browse/KAFKA-7523

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

pgwhalen committed:
Add ConnectedStoreProvider to Processor API

 - Implementing KIP-401, ConnectedStoreProvider is now the
 superinterface ofvProcessor/TransformerSuppliers, allowing the user to
 provide state stores that will automatically be added to the topology
 and connected to the associated processor or transformer.  This is an
 alternate way to connect state stores to a processor, and cannot be
 used along with passing the varags stateStoreNames to
 KStream::process/transform. It also works with Topology::addProcessor.
 - WordCountTransformerDemo is a middle ground between the low level
 Processor API and the high level DSL, using KStream::transform and
 ConnectedStoreProvider
 - The restriction on adding the same state store twice is relaxed so
 that multiple ConnectedStoreProviders can add the same store.

pgwhalen committed:
Fix typo in stream map

 - need to get the name from the StoreBuilder as opposed building it

pgwhalen committed:
Ensure delegation to ConnectedStoreProvider#stores

 - When TransformerSuppliers are wrapped in ProcessorSuppliers
 internally, supplier.stores() must be propagated up from the internal
 supplier to the wrapping supplier.

pgwhalen committed:
Remove restriction on connecting stores

 - Using both stateStoreNames and ConnectedStoreProvider is permitted.

pgwhalen committed:
Merge branch 'trunk' into kip-401-store-provider

# Conflicts:
#	streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
#	streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

pgwhalen committed:
Improve missing store message

pgwhalen committed:
Test TransformerSupplierAdapter store passthrough

pgwhalen committed:
Make store name list mutable

pgwhalen committed:
Fix TopologyTest

pgwhalen committed:
Add multiple processors/one store test

pgwhalen committed:
Merge branch 'trunk' into kip-401-store-provider

pgwhalen committed:
Fix dupe state store test

pgwhalen committed:
Fix ValueTransformer* delegation

pgwhalen committed:
Fix KStreamTransform test

pgwhalen committed:
Handle null varargs

pgwhalen committed:
Get ValueTransformer each time

pgwhalen committed:
Update site docs

pgwhalen commented:
Requesting a review from @mjsax @bbejeck @guozhangwang since they were binding voters on the KIP (though of course anyone is welcome to review).

In my view this is just about complete, though I have a few minor open questions:
- How should I go about updating the `KStream.scala` Scaladocs?  They're considerably smaller and therefore don't seem to demand fully copying from `KStream.java`, but even the small ones are slightly incorrect now that there are multiple ways to add and connect state stores.  My best guess is to just come up with a similarly minified version for `KStream.scala`.
- Is there some rule of thumb or tooling for how to write the HTML for kafka website?  I'm no front end guy, and all of the `span`s with different `class`es used in the code blocks have me thoroughly confused.  It's so opaque that it suggests to me that I'm missing out on some tooling to generate this HTML.  Specifically, this line https://github.com/apache/kafka/pull/6824/files#diff-1d9b6a3809b5c863a78cb3ec1773d106R423 is what presumably needs more `span`s.  For readability I would prefer to make that multiple lines, but before going through the agony of writing that HTML I figured I would ask if there was an easier way.  I will say that [this page](https://cwiki.apache.org/confluence/display/KAFKA/Setup+Kafka+Website+on+Local+Apache+Server) was super helpful though, thanks @guozhangwang .
- I still have to duplicate some of my `KStream.java` Javadoc changes, but I wanted to get this out for review first; if there are changes to the copied portions it will be less laborious to reduplicate.

pgwhalen commented:
Pinging @mjsax @bbejeck @guozhangwang again since 2.4.0 deadlines are approaching.  Would be good to get some high level guidance at least so I can set aside time for larger changes.

mjsax commented:
@pgwhalen Sorry that this PR slips for `2.4` release -- it was too much other work and we did not find time to review on time... :(

Hope you are not too disappointed. I'll try to review soon and it should be not problem at all to get it into 2.5.

pgwhalen commented:
No worries, I saw lots of really great stuff making it in at the last minute, totally makes sense. 

> On Oct 7, 2019, at 1:26 PM, Matthias J. Sax <notifications@github.com> wrote:
> 
> @pgwhalen Sorry that this PR slips for 2.4 release -- it was too much other work and we did not find time to review on time... :(
> 
> Hope you are not too disappointed. I'll try to review soon and it should be not problem at all to get it into 2.5.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

mumrah commented:
Besides needing a rebase with trunk, what is the status of this PR? Do we want to pull it into the 2.5 release?

pgwhalen commented:
I’d like to pull it in if possible. Still needs a review obviously, a bit of additional documentation,  and the rebase isn’t totally trivial (I will have to add documentation to some methods that were created since the PR was created), but I can get going on that soon. 

> On Jan 21, 2020, at 2:38 PM, David Arthur <notifications@github.com> wrote:
> 
> ﻿
> Besides needing a rebase with trunk, what is the status of this PR? Do we want to pull it into the 2.5 release?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

mjsax commented:
The 2.4 release was quite challenging and heavily delayed -- hence, we did not really got time to continue to review this PR yet :( (the 2.5 release cycle was effectively super short for us).

> it should be not problem at all to get it into 2.5.

Last famous words... :(

Atm, it's quite tight for us to get KIP-447 and KIP-535 (plus KIP-562 that is a small change to KIP-535) into 2.5. Furthermore, KIP-446 and KIP-221 are also in the pipeline and slipped 2.4 release similar to KIP-401 already. We also just voted KIP-216 and KIP-399 that might both slip 2.5.

What I am trying to say: it's great that we have so many KIPs but it's super hard for us to keep up. Maybe @bbejeck could carve out some cycles to allow us getting it into 2.4 -- I personally won't have time before feature freeze to review. I am really sorry. :(

SoerenHenning commented:
What is the current status on this? It there any chance to get this merged for the 2.6.0 release as mentioned in the [release plan](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=152113430)? For one of our implementations, this KIP is a huge API enhancement.

guozhangwang commented:
The KIP is accepted so I think we'd look into the PR authored by Paul,
@Matthias
J Sax <matthias@confluent.io> was reviewing it so maybe you can tell us the
latest status?


Guozhang

On Mon, May 11, 2020 at 6:53 AM Sören Henning <notifications@github.com>
wrote:

> What is the current status on this? It there any chance to get this merged
> for the 2.6.0 release as mentioned in the release plan
> <https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=152113430>?
> For one of our implementations, this KIP is a huge API enhancement.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/kafka/pull/6824#issuecomment-626715421>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACR3Q7LODPS3FTTIRRR2NGDRQ77M7ANCNFSM4HP53E6Q>
> .
>


-- 
-- Guozhang

mjsax commented:
KIP-221 was merged recently and I was considering to pick up the review for PR for KIP-216 next (because it has a longer wait time already). But I guess if there is more demand here, we could also do this PR before KIP-216...

@pgwhalen What is your current status. If we review this PR this week, can you work on addressing the comment so we can get it into 2.6 release? Feature freeze deadline in 5/27.

pgwhalen commented:
Yep, I'd be happy to pick it back up in short order.  It's obviously a bit stale, so I think it'll take me a few days to find time to properly clean up the conflicts with `trunk`.  But looking at the conflicts I think it will mostly be cosmetic cleanup, so it's ready for some eyes already at least.

mjsax commented on docs/streams/developer-guide/dsl-api.html:
```diff
@@ -3164,10 +3164,17 @@ <h4><a id="streams_concepts_globalktable" href="#streams_concepts_globalktable">
+                        <li>By passing the name of a store that has already been added via <code class="docutils literal"><span class="pre">Topology#addStateStore()</span></code> to the corresponding <code class="docutils literal"><span class="pre">KStream#process()</span></code> method call.</li>
+                        <li>Implementing <code class="docutils literal"><span class="pre">ConnectedStoreProvider#stores()</span></code> on the <code class="docutils literal"><span class="pre">ProcessorSupplier</span></code>
+                            passed to <code class="docutils literal"><span class="pre">KStream#process()</span></code>.  In this case there is no need to call <code class="docutils literal"><span class="pre">Topology#addStateStore()</span></code>
+                            beforehand, the store will be automatically added for you.
```
nit: `,` -> `;` ?

mjsax commented on docs/streams/developer-guide/processor-api.html:
```diff
@@ -396,33 +396,52 @@ <h2><a class="toc-backref" href="#id8">Connecting Processors and State Stores</a
+                        runtime, indicating the state store is not accessible from this processor.</p>
+                    <p>Now that you have fully defined your processor topology in your application, you can proceed to
+                        <a class="reference internal" href="running-app.html#streams-developer-guide-execution"><span class="std std-ref">running the Kafka Streams application</span></a>.</p>
+                </div>
```
Meta comment: can you also extend `streams/upgrade-guide.html` -- there is a section "public API" changes for the 2.6 release.

mjsax commented on docs/streams/developer-guide/processor-api.html:
```diff
@@ -396,33 +396,52 @@ <h2><a class="toc-backref" href="#id8">Connecting Processors and State Stores</a
+                        to dynamically determine the Kafka topic to write to for each received record from the upstream processor).</li>
+                </ul>
+                <p>
+                    In some cases, it may be more convenient to add and connect a state store at the same time as you add the processor it is connected to to the topology.
```
Not sure what this means: `it is connected to to the topology.`

mjsax commented on docs/streams/developer-guide/processor-api.html:
```diff
@@ -396,33 +396,52 @@ <h2><a class="toc-backref" href="#id8">Connecting Processors and State Stores</a
+                <p>
+                    In some cases, it may be more convenient to add and connect a state store at the same time as you add the processor it is connected to to the topology.
+                    This can be done by implementing <code class="docutils literal"><span class="pre">ConnectedStoreProvider#stores()</span></code> on the <code class="docutils literal"><span class="pre">ProcessorSupplier</span></code>
+                    in place of calling <code class="docutils literal"><span class="pre">Topology#addStateStore()</span></code>, like this:
```
nit: `in place` -> `instead` ?

mjsax commented on docs/streams/developer-guide/processor-api.html:
```diff
@@ -396,33 +396,52 @@ <h2><a class="toc-backref" href="#id8">Connecting Processors and State Stores</a
+<span class="n">builder</span><span class="o">.</span><span class="na">addSource</span><span class="o">(</span><span class="s">&quot;Source&quot;</span><span class="o">,</span> <span class="s">&quot;source-topic&quot;</span><span class="o">)</span>
+
+    <span class="c1">// add the WordCountProcessor node which takes the source processor as its upstream processor, along with its state store</span>
+    <span class="o">builder.addProcessor("Process", new ProcessorSupplier() { public Processor&lt;String, String&gt; get() { return new WordCountProcessor(); } public Set&lt;StoreBuilder&gt; stores() { return countStoreBuilder; } });</span>
```
Can we put this code (ie, the definition of `ProcessorSupplier`) into multiple lines?
```
builder.addProcessor("Process", new ProcessorSupplier() {
    public Processor&lt;String, String&gt; get() {
        return new WordCountProcessor();
    }
    public Set&lt;StoreBuilder&gt; stores() { 
        return countStoreBuilder;
    }
});
```

Also note that in the original example above, there is an indention because method calls are chained:
```
builder.source()
    .process()
    .addStateStore()
    .addSink();
```

Either chain _all_ method calls, too, or chain none. If you don't chain, remove the indention. Currently it would render as follows (including an incorrect `;` after `addProcessor()`:
```
builder.addSource();
    builder.addProcessor();
    .addSink();
```

mjsax commented on docs/streams/developer-guide/processor-api.html:
```diff
@@ -396,33 +396,52 @@ <h2><a class="toc-backref" href="#id8">Connecting Processors and State Stores</a
+    <span class="o">.</span><span class="na">addSink</span><span class="o">(</span><span class="s">&quot;Sink&quot;</span><span class="o">,</span> <span class="s">&quot;sink-topic&quot;</span><span class="o">,</span> <span class="s">&quot;Process&quot;</span><span class="o">);</span>
+</pre></div>
+                    <p>This allows for a processor to "own" state stores, effectively encapsulating its usage from the user wiring the topology.
+                        Multiple processors that share a state store may provide the same store with this technique, as long as the <code class="docutils literal"><span class="pre">StoreBuilder</span></code> is the same instance.</p>
```
nit: highlight `instance`

mjsax commented on docs/streams/developer-guide/processor-api.html:
```diff
@@ -396,33 +396,52 @@ <h2><a class="toc-backref" href="#id8">Connecting Processors and State Stores</a
+                        <a class="reference internal" href="running-app.html#streams-developer-guide-execution"><span class="std std-ref">running the Kafka Streams application</span></a>.</p>
+                </div>
             </div>
-            <p>Here is a quick explanation of this example:</p>
```
I seems you did not really change anything here? (Only the indention? Why? -- Make it hard to review.)

mjsax commented on streams/examples/src/main/java/org/apache/kafka/streams/examples/wordcount/WordCountTransformerDemo.java:
```diff
@@ -0,0 +1,159 @@
+import java.util.concurrent.CountDownLatch;
+
+/**
+ * Demonstrates, using a {@link Transformer} which combines the low-level Processor APIs with the high-level KStream DSL,
```
`KStream` -> `Kafka Streams`

(`KStream` is not an appropriate abbreviation for Kafka Streams, because the DSL uses the abstraction of `KStream` and `KTable` and thus it would be ambiguous.)

mjsax commented on streams/examples/src/main/java/org/apache/kafka/streams/examples/wordcount/WordCountTransformerDemo.java:
```diff
@@ -0,0 +1,159 @@
+ * represent lines of text; and the histogram output is written to topic "streams-wordcount-processor-output" where each record
+ * is an updated count of a single word.
+ * <p>
+ * This example differs from {@link WordCountProcessorDemo} in that it uses a {@link Transformer} and {@link ConnectedStoreProvider}
```
`uses a {@link Transformer} and {@link ConnectedStoreProvider}` -- this might be confusing because `TransformerSupplier extends ConnectedStoreProvider` and the code does not contain any direct reference to `ConnectedStoreProvider`... Can we rephrase it?

mjsax commented on streams/examples/src/main/java/org/apache/kafka/streams/examples/wordcount/WordCountTransformerDemo.java:
```diff
@@ -0,0 +1,159 @@
+                        }
+                    }
+
+                    context.commit();
```
I think we should remove `context.commit();` (Can we also remove it in `WordCountProcessorDemo`?)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -885,49 +887,71 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
      * <p>
-     * In order to assign a state, the state must be created and registered beforehand (it's not required to connect
-     * global state stores; read-only access to global state stores is available by default):
-     * <pre>{@code
```
This line must be kept to start the code example

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -885,49 +887,71 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
      * builder.addStateStore(keyValueStoreBuilder);
      *
-     * KStream outputStream = inputStream.transform(new TransformerSupplier() { ... }, "myTransformState");
+     * KStream outputStream = inputStream.transform(new TransformerSupplier() { public Transformer get() { return new MyTransformer(); } }, "myTransformState");
```
Should be multiple lines to improve readabilty.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -953,7 +977,7 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
      *
      * @param transformerSupplier an instance of {@link TransformerSupplier} that generates a {@link Transformer}
-     * @param stateStoreNames     the names of the state stores used by the processor
+     * @param stateStoreNames     the names of the state stores used by the transformer, passed only if {@link ConnectedStoreProvider#stores()} is null
```
Why this:
`, passed only if {@link ConnectedStoreProvider#stores()} is null` 

From the KIP: 
> A user may continue to "connect" stores to a processor by passing stateStoreNames when calling stream.process/transform(...) .  This may be used in combination with a Supplier  that provides its own state stores by implementing ConnectedStoreProvider::stores() .

pgwhalen replied:
Good catch, I will change this with the revised `KStream` javadocs coming soon.  It was a relic of my original proposal; the code was written before we decided to change it and I missed removing this.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -1162,53 +1186,77 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
      * builder.addStateStore(keyValueStoreBuilder);
      *
-     * KStream outputStream = inputStream.flatTransform(new TransformerSupplier() { ... }, "myTransformState");
+     * KStream outputStream = inputStream.flatTransform(new TransformerSupplier() { public Transformer get() { return new MyTransformer(); } }, "myTransformState");
```
as above (more below; not add comments again)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -885,49 +887,71 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
+     *         return new MyTransformer();
+     *     }
+     *
+     *     // provide store(s) that will be added and connected to the associated transformer
```
Should we add a comment to highlight that the store name `"myTransformState"` from the builder is used to access the store via the `context` later?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -1231,7 +1279,7 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
      * @param transformerSupplier an instance of {@link TransformerSupplier} that generates a {@link Transformer}
      * @param named               a {@link Named} config used to name the processor in the topology
-     * @param stateStoreNames     the names of the state stores used by the processor
+     * @param stateStoreNames     the names of the state stores used by the transformer, passed only if {@link ConnectedStoreProvider#stores()} is null
```
As above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -1915,43 +2061,66 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
-     *             }
-     *         }
+     *     void close() {
+     *         // can access this.state
```
`process()` has an overload below that needs and update, too. There are also some `transform()` variants you missed.

Overall, it should be 14 methods that needs to be updated. I counted only 8 that this PR updates.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -700,12 +703,29 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
+     * 2) {@link StoreBuilder}s are provided by the Processor/TransformerSupplier itself, by returning a set from
+     * {@link ConnectedStoreProvider#stores()}.  The {@link StoreBuilder}s will also be added to the topology.
+     */
+    private String[] getStoreNamesAndMaybeAddStores(final ConnectedStoreProvider storeProvider, final String[] varargsStoreNames) {
```
Why not make `varargsStoreNames` a vararg? The `null` check below is not required then?

Why do we actually "unwrap" the stores here and add them one-by-one to the builder/topology?

pgwhalen replied:
>Why not make varargsStoreNames a vararg?

Good point, will change.

>Why do we actually "unwrap" the stores here and add them one-by-one to the builder/topology?

I'm not sure I understand what you mean by "unwrap" and "add one-by-one."  Are you saying I should add a method `InternalStreamsBuilder::addStateStores` that takes a `Collection`/`Set`?

mjsax replied:
Sorry for not being detailed enough. My bad (re-reading my comment it does not really make sense). What I actually meant was:

Within `KStreamImpl` we should not access the `builder` at all. Kafka Streams uses a 2-phase translation: in the first phase, we only build up a logical representation of the topology (what we call the "stream graph" that comprises "stream graph node"). If optimization is enabled, the optimizer would potentially modify the "stream graph" before it gets translated into the actual topology.

The created `StatefulProcessorNode` is one of those "stream graph nodes" -- in its `writeToTopology()` method it will add it's state store to the `Topology` and connect the processor to other stores if necessary.

Thus, your code adds the store "too early" and returns a list of store names that the processor is later connecting to. Instead, we should modify `StatefulProcessorNode` and delay adding the store to the topology until `writeToTopology()` is called: in particular `StatefulProcessorNode` should have a constructor that accepts both an array of store names it should connect to and a `ConnectedStoreProvider`.

Does this clarify?

pgwhalen replied:
Makes total sense, apologies for missing the subtlety.  It's possible I wrote the first version of this code before there even was that extra optimization step, and never bothered to change my mental model of how it should work.

mjsax replied:
No worries. All good. :)

Btw: the optimization layer was added in 2.0

mjsax commented on streams/src/test/java/org/apache/kafka/streams/TopologyTest.java:
```diff
@@ -291,13 +291,19 @@ private void mockStoreBuilder() {
 
     @Test
-    public void shouldNotAllowToAddStoreWithSameName() {
+    public void shouldNotAllowToAddStoreWithSameNameAndDifferentInstance() {
```
We should add a new test `shouldAllowToShareStoreUsingSameStoreBuilder`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KStreamTransformIntegrationTest.java:
```diff
@@ -55,61 +60,92 @@
-                                                                       new KeyValue<>(2, 1),
-                                                                       new KeyValue<>(2, 3),
-                                                                       new KeyValue<>(1, 3))));
+                new KeyValue<>(2, 2),
```
If you change the indention, move the first `new KeyValue(1, 1)` to it's own line.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilderTest.java:
```diff
@@ -329,10 +329,11 @@ public void testAddStateStoreWithSink() {
 
     @Test
-    public void testAddStateStoreWithDuplicates() {
+    public void testAddStateStoreWithDifferentInstances() {
```
Can we improve the test name -> `shouldNotAllowToAddStoresWithSameName` (or similar)

We should also add a test that verifies that adding the same `StoreBuilder` instance multiple times works.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:
```diff
@@ -549,6 +593,14 @@ private Topology createStatefulTopology(final String storeName) {
     }
 
+    private Topology createConnectedStateStoreTopology(final String storeName) {
+        final StoreBuilder<KeyValueStore<String, String>> storeBuilder = Stores.keyValueStoreBuilder(Stores.inMemoryKeyValueStore(storeName), Serdes.String(), Serdes.String());
```
nit: line too long

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:
```diff
@@ -723,6 +775,20 @@ public void process(final String key, final String value) {
+        return new ProcessorSupplier<K, V>() {
+            @Override
+            public Processor<K, V> get() {
+                return processor;
```
This does not seem to be a good implementation? A supplier is supposed to return a new instance on each `get()` call. It might not be an issue now, but might be error prone in the future. 

Maybe accept an supplier as input instead of a `Processor`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:
```diff
@@ -213,6 +215,48 @@ public void testDrivingStatefulTopology() {
+            .addSource("source1", STRING_DESERIALIZER, STRING_DESERIALIZER, INPUT_TOPIC_1)
+            .addSource("source2", STRING_DESERIALIZER, STRING_DESERIALIZER, INPUT_TOPIC_2)
+            .addProcessor("processor1", defineWithStores(new StatefulProcessor(storeName), Collections.singleton(storeBuilder)), "source1")
+            .addProcessor("processor2", defineWithStores(new StatefulProcessor(storeName), Collections.singleton(storeBuilder)), "source2")
```
The same `StatefulProcessor` instance would be used in each processor step -- I would expect this to be problematic (cf. my comment below on `defineWithStores()`)

pgwhalen replied:
I'm a little confused - the code `new`s different `StatefulProcessor`s for each step, so I don't see how it would be reused.

It would certainly be a risk if the result of `defineWithStores` is reused, but in practice it is not; it's just a sort of macro for defining tests.  I was just trying to emulate the existing `define`, which has the same issue.  Perhaps I should make your suggested change as well as remove `define()` entirely?

mjsax replied:
Well, `get()` of the returned supplier could be called multiple times, thus, would return the name object that is passed into `defineWithStores` each time -- this is incorrect though, because each time `get()` is called a new object must be returned.

pgwhalen replied:
Makes sense.  Would you like me to change `define()` as well, since it also creates a supplier that returns the same object each time?

mjsax replied:
I think the simplest fix is, to pass a Java `supplier` a first argument to `define()`:
```
defineWithStores(() -> new StatefulProcessor(storeName), Collections.singleton(storeBuilder))
```

And within `define()`:
```
public Processor<K, V> get() {
    return supplier.get();
}
```

mjsax reviewed (commented):
Made an initial pass.

pgwhalen committed:
Merge branch 'trunk' into kip-401-store-provider

 - handles conflicts except for KStream.java which will be rewritten for a future commit

# Conflicts:
#	streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
#	streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
#	streams/src/test/java/org/apache/kafka/streams/integration/KStreamTransformIntegrationTest.java
#	streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java

pgwhalen committed:
Split long line

pgwhalen committed:
Add test for adding StoreBuilder twice to builder

 - rename another test

pgwhalen committed:
Fix indentation

pgwhalen committed:
Add @see's to ConnectedStoreProvider

pgwhalen committed:
Add shouldAllowToShareStoreUsingSameStoreBuilder

pgwhalen committed:
Make varargsStoreNames varargs

pgwhalen committed:
Clarify WordCountTransformerDemo javadoc

pgwhalen committed:
Punctuation nit

pgwhalen committed:
Remove context.commit()

pgwhalen commented:
@mjsax Thanks for the initial review!

I believe I have addressed all comments (some of which I had questions about) except for comments on documentation in `processor-api.html` and `KStream.java` specifically.  For the latter I just abandoned my changes in the merge from `trunk` for now; I will re-apply them within the next couple of days with the extra variants created since this PR was initially created (and the changes you've suggested).  Just didn't have time to get to them tonight.

I asked this on the dev mailing list a while ago but didn't get a great answer - is there some sort of tooling I can use to edit the HTML docs like `processor-api.html`?  I really don't understand what all the HTML classes are (e.g. `c1`, `o`, `na`, `s` etc.), and can't figure out how to get rendered code formatted the way I want without extraordinarily tedious trial and error; that's one of the reasons the rendered code is formatted poorly.

pgwhalen committed:
Delay adding stores until writeToTopology

pgwhalen committed:
Fix defineWithStores

pgwhalen committed:
upgrade-guide.html note

pgwhalen commented:
Retest this please.

mjsax commented:
> I asked this on the dev mailing list a while ago but didn't get a great answer - is there some sort of tooling I can use to edit the HTML docs like processor-api.html? I really don't understand what all the HTML classes are (e.g. c1, o, na, s etc.), and can't figure out how to get rendered code formatted the way I want without extraordinarily tedious trial and error; that's one of the reasons the rendered code is formatted poorly.

This is a real pain! I have no idea what tool you could use. Originally, we only have very simple HTML, and at some point we ended up with this "mess"... Sorry that I don't have a better answer. I also "fight" with it when I do a PR and have to edit the HTML manually... To render the webpage, you can setup a local Apache WebServer (cf https://cwiki.apache.org/confluence/display/KAFKA/Contributing+Website+Documentation+Changes) -- maybe that helps.

Btw: only committers can trigger Jenkins jobs... (this was changes a while ago from ASF Infra team to "guard" the test infrastructure).

mjsax commented:
Retest this please.

mjsax commented:
Retest this please.

pgwhalen committed:
Add KStream documentation

pgwhalen committed:
Merge branch 'trunk' into kip-401-store-provider

# Conflicts:
#	streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java

pgwhalen committed:
Fix processor-api docs

pgwhalen committed:
Fix code formatting

 - Use the better way of formatting code
 for the other "connecting processors and state stores"
 example

pgwhalen commented:
@mjsax I believe I have made every change requested and this is ready for a final review/approval.

Regarding the tedious HTML editing, I noticed that a lot of documentation uses `<pre class="brush: java">` which makes it far easy to get good formatting, so I just switched to using that.  I took the liberty of changing the other code example in the same section to use the same method, even though I didn't change the content.  I can revert that if you'd like, I just figured it would look nicer.

I think it would be worth standardizing code examples across all the documentation to avoid pain/confusion like this for others in the future.  Obviously this sort of cleanup isn't at the top of the committers' list of things to do, but I'd be happy to take a stab at it myself or at least create a ticket if you think it's a good idea.

pgwhalen commented:
Whoops, sent that just a bit early; I have a few failing tests still.

pgwhalen committed:
Fix build/tests

 - final
 - testDrivingConnectedStateStoreInDifferentProcessorsTopology
 - Remove addStateStore from KStreamTransformIntegrationTest @before
 since it is only needed in half the tests

pgwhalen commented:
Okay, looks good for real now.

I can't run the `KStreamTransformIntegrationTest` because of [this issue](https://github.com/confluentinc/ksql/issues/1332) (despite trying many ways to resolve it), but the fact that the tests all get as far as that failure is a good sign at least, and gives me confidence that my change is okay.

vvcephei commented:
Test this please

vvcephei commented:
Test this please

vvcephei commented:
Test this please

vvcephei commented:
Retest this please

vvcephei commented:
Heh, I guess Jenkins is still on vacation. I'll try again later.

vvcephei commented on docs/streams/developer-guide/processor-api.html:
```diff
@@ -381,22 +381,18 @@ <h2><a class="toc-backref" href="#id8">Connecting Processors and State Stores</a
-    <span class="o">.</span><span class="na">addSink</span><span class="o">(</span><span class="s">&quot;Sink&quot;</span><span class="o">,</span> <span class="s">&quot;sink-topic&quot;</span><span class="o">,</span> <span class="s">&quot;Process&quot;</span><span class="o">);</span>
-</pre></div>
-            </div>
+            <pre class="brush: java">
```
Thanks for the cleanup here. Do you mind posting screenshots of the rendered docs?

pgwhalen replied:
After the `"Source"` addition suggested in another comment:

![screencapture-localhost-25-documentation-streams-developer-guide-processor-api-html-2020-05-26-13_27_24](https://user-images.githubusercontent.com/3172488/82937232-6e501500-9f55-11ea-96ed-26a2a4c0457c.png)

vvcephei replied:
Thanks!

vvcephei commented on docs/streams/developer-guide/processor-api.html:
```diff
@@ -409,7 +405,31 @@ <h2><a class="toc-backref" href="#id8">Connecting Processors and State Stores</a
+                        public Set&ltStoreBuilder&lt?&gt&gt stores() {
+                            return countStoreBuilder;
+                        }
+                    })
```
Does this need a third arg, "Source"?

pgwhalen replied:
Good catch! Will add

vvcephei commented on streams/examples/src/main/java/org/apache/kafka/streams/examples/wordcount/WordCountProcessorDemo.java:
```diff
@@ -93,8 +93,6 @@ public void process(final String dummy, final String line) {
                         }
                     }
-
-                    context.commit();
```
Sorry if I've missed something in the discussion history. Why does this need to come out?

pgwhalen replied:
https://github.com/apache/kafka/pull/6824#discussion_r424769188

If I were to guess, @mjsax wanted it removed because it isn't necessary and distracts from the most simple form of the example.

vvcephei replied:
Ah, I _did_ miss something. Thanks.

vvcephei commented on streams/examples/src/main/java/org/apache/kafka/streams/examples/wordcount/WordCountTransformerDemo.java:
```diff
@@ -0,0 +1,160 @@
+ * {@code bin/kafka-topics.sh --create ...}), and write some data to the input topic (e.g. via
+ * {@code bin/kafka-console-producer.sh}). Otherwise you won't see any data arriving in the output topic.
+ */
+public final class WordCountTransformerDemo {
```
Should there be an integration test for this (to make sure it actually continues to work in the face of future changes)?

pgwhalen replied:
Great point, I'll add. I did not realize there was one for the class I was modeling (`WordCountProcessorDemo`)

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/integration/KStreamTransformIntegrationTest.java:
```diff
@@ -55,61 +60,93 @@
     private final String topic = "stream";
     private final String stateStoreName = "myTransformState";
     private final List<KeyValue<Integer, Integer>> results = new ArrayList<>();
-    private final ForeachAction<Integer, Integer> action = (key, value) -> results.add(KeyValue.pair(key, value));
+    private final ForeachAction<Integer, Integer> accumulateExpected = (key, value) -> results.add(KeyValue.pair(key, value));
```
Not your fault, but I couldn't figure out how this test actually works. It doesn't look like we ever clear `results` between runs. What's going on here?

pgwhalen replied:
I would assume it's a JUnit test that uses the [standard lifecycle](https://stackoverflow.com/a/19381563/1031507) where the class is recreated for every test.  I haven't looked into the test running infrastructure at all though so maybe because this is `@Category({IntegrationTest.class})` you would expect some other behavior?

vvcephei replied:
Huh, I never knew that. After reading into it a little more, I can see why it works, but it's also not guaranteed to work for all Runner implementations; the docs just say this is the behavior of the _default_ implementation.

Anyway, like I said, it predates your PR, so I wouldn't request any changes. I just thought you might have some insight since you had modified the test. And you did! ;)

Thanks.

vvcephei reviewed (commented):
Hey @pgwhalen , thanks for this! I've just completed a pass, and everything looks good to me. I just left a few comments.

pgwhalen committed:
Add "Source"

pgwhalen committed:
Fix/Add Demo tests

 - WordCountTransformerTest tests the example
 WordCountTransformerDemo
 - WordCountProcessorTest should not check if the
 processor committed anymore

pgwhalen commented:
@vvcephei I made your suggested changes and fixed the broken test (which was related to https://github.com/apache/kafka/pull/6824#discussion_r430522689)

vvcephei commented:
Test this please

vvcephei commented:
Test this please

vvcephei commented:
Ok to test

vvcephei reviewed (approved):
For my part, this looks good. Thanks @pgwhalen !

vvcephei commented:
Style error:
```
00:08:09.499 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk14-scala2.13/streams/examples/src/test/java/org/apache/kafka/streams/examples/wordcount/WordCountTransformerTest.java:28: Using the '.*' form of import should be avoided - org.junit.Assert.*. [AvoidStarImport]
```

pgwhalen committed:
Remove * import

pgwhalen commented:
Shoot, sorry, I thought I had managed to stop IntelliJ from doing that.  Just pushed the fix.

mjsax commented:
Retest this please.

mjsax commented:
Retest this please.

mjsax commented:
Retest this please.

mjsax commented on docs/streams/developer-guide/dsl-api.html:
```diff
@@ -3533,10 +3533,17 @@ <h5><a class="toc-backref" href="#id34">KTable-KTable Foreign-Key
+                    <ul class="simple">
+                        <li>By passing the name of a store that has already been added via <code class="docutils literal"><span class="pre">Topology#addStateStore()</span></code> to the corresponding <code class="docutils literal"><span class="pre">KStream#process()</span></code> method call.</li>
+                        <li>Implementing <code class="docutils literal"><span class="pre">ConnectedStoreProvider#stores()</span></code> on the <code class="docutils literal"><span class="pre">ProcessorSupplier</span></code>
+                            passed to <code class="docutils literal"><span class="pre">KStream#process()</span></code>.  In this case there is no need to call <code class="docutils literal"><span class="pre">Topology#addStateStore()</span></code>
```
```suggestion
                            passed to <code class="docutils literal"><span class="pre">KStream#process()</span></code>.  In this case there is no need to call <code class="docutils literal"><span class="pre">StreamsBuilder#addStateStore()</span></code>
```

mjsax replied:
This is the DSL guide, so using `StreamsBuilder` seems appropriate.

pgwhalen committed:
StreamsBuilder instead of Topology

Co-authored-by: Matthias J. Sax <mjsax@apache.org>

mjsax commented on streams/src/main/java/org/apache/kafka/streams/Topology.java:
```diff
@@ -656,6 +660,14 @@ public synchronized Topology addProcessor(final String name,
+                internalTopologyBuilder.addStateStore(storeBuilder, name);
+            }
+            final String[] storeNames = stores.stream().map(StoreBuilder::name).toArray(String[]::new);
+            internalTopologyBuilder.connectProcessorAndStateStores(name, storeNames);
```
Why do we need this? Calling `addStateStore(storeBuilder, name)` should connect the store to the processor already?

pgwhalen replied:
Good point, removing.

mjsax commented on docs/streams/developer-guide/dsl-api.html:
```diff
@@ -3533,10 +3533,17 @@ <h5><a class="toc-backref" href="#id34">KTable-KTable Foreign-Key
+                        <li>Implementing <code class="docutils literal"><span class="pre">ConnectedStoreProvider#stores()</span></code> on the <code class="docutils literal"><span class="pre">ProcessorSupplier</span></code>
+                            passed to <code class="docutils literal"><span class="pre">KStream#process()</span></code>.  In this case there is no need to call <code class="docutils literal"><span class="pre">StreamsBuilder#addStateStore()</span></code>
+                            beforehand; the store will be automatically added for you.
+                        </li>
```
Should we add a sentence, that `ConnectedStoreProvided` can also be used for `TransformerSupplier` et al.

mjsax reviewed (approved):
LGMT. Thanks for the PR and for patients! Took way to long to get this in.

Btw: Outstanding job on updating the JavaDocs!

(Just need to convince Jenkins to retest this. Can merge afterwards.)

pgwhalen committed:
Remove unnecessary adding to store

pgwhalen committed:
Add details on ConnectedStoreProvider variants

pgwhalen commented:
I made the final suggestions, so we should be good to merge on passing tests.

Thank you for still having the appetite to include even though it took a while!  For my next contribution, I'll be more proactive on getting a reviewable PR when it's _not_ crunch time. 😄

mjsax commented:
Retest this please.

mjsax commented:
Retest this please.

vvcephei commented:
I don't think any of those failures are related:
    org.apache.kafka.connect.integration.InternalTopicsIntegrationTest.testCreateInternalTopicsWithFewerReplicasThanBrokers

org.apache.kafka.streams.integration.EosBetaUpgradeIntegrationTest.shouldUpgradeFromEosAlphaToEosBeta[true]

org.apache.kafka.streams.integration.EosBetaUpgradeIntegrationTest.shouldUpgradeFromEosAlphaToEosBeta[false]

vvcephei commented:
Retest this please

mjsax commented:
Java 8: org.apache.kafka.streams.integration.EosBetaUpgradeIntegrationTest.shouldUpgradeFromEosAlphaToEosBeta[false]

Java 11 and 14 passed. Merging this.

mjsax commented:
Thanks for the KIP and PR @pgwhalen!

vvcephei commented:
Yes, thanks so much for this contribution, and your indefatigable patience, @pgwhalen !

SoerenHenning commented:
Thanks for implementing this @pgwhalen and thanks for merging @mjsax @vvcephei as well. I'm really looking forward to refactoring my code when this is released!
